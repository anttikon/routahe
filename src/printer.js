import { blueBright, bold, cyan, gray, green, magenta, yellow } from 'chalk'
import moment from 'moment'
import Table from 'cli-table/lib'
import tableStyle from './tableStyle'

export const printLocations = (args) => {
  console.log(bold(`${args.from.label} ${gray('->')} ${args.to.label}`))
  return args
}

export const printHelp = () => {
  console.log('  Examples:')
  console.log('')
  console.log(gray('    Default from - to usage'))
  console.log('    $ routahe kamppi pasila')
  console.log('')
  console.log(gray('    Specify departure time'))
  console.log('    $ routahe kamppi pasila 5:30')
  console.log('')
  console.log(gray('    Specify arrival time with @'))
  console.log('    $ routahe kamppi pasila @12:30')
  console.log('')
  console.log(gray('    Specify date with time'))
  console.log('    $ routahe kamppi pasila 12:30 24.12.')
  console.log('')
}

const getRowProps = mode => {
  switch (mode) {
    case 'WALK':
      return { emoji: String.fromCodePoint(0x1f6b6), color: gray, name: mode }
    case 'RAIL':
      return { emoji: String.fromCodePoint(0x1f685), color: magenta, name: mode }
    case 'BUS':
      return { emoji: String.fromCodePoint(0x1f68c), color: blueBright, name: mode }
    case 'TRAM':
      return { emoji: String.fromCodePoint(0x1f683), color: green, name: mode }
    case 'SUBWAY':
      return { emoji: String.fromCodePoint(0x1f687), color: yellow, name: 'METRO' }
    case 'FERRY':
      return { emoji: String.fromCodePoint(0x1f6a2), color: cyan, name: mode }
    default:
      return { emoji: String.fromCodePoint(0x1f937), color: value => value, name: mode }
  }
}

const getMinAttribute = (array = [], field) => array.reduce((acc, obj) => !acc || acc > obj[field] ? obj[field] : acc, undefined)
const getMaxAttribute = (array = [], field) => array.reduce((acc, obj) => !acc || acc < obj[field] ? obj[field] : acc, undefined)

const getRouteHeaderRow = route => {
  const startTime = moment(getMinAttribute(route.legs, 'startTime'))
  const endTime = moment(getMaxAttribute(route.legs, 'endTime'))

  return `\n${startTime.format('HH:mm')} - ${endTime.format('HH:mm')} ${endTime.diff(startTime, 'minutes')}min`
}

const getRouteLegRow = leg => {
  const { color, emoji, name } = getRowProps(leg.mode)
  return [
    color('  |'),
    `${moment(leg.startTime).format('HH:mm')} - ${moment(leg.endTime).format('HH:mm')}`,
    `${emoji} ${color(name)}`,
    leg?.route?.shortName || '',
    '->',
    leg.to.name,
  ]
}

const stringsToTable = strings => {
  const table = new Table(tableStyle)
  table.push(...strings)
  return table
}

const getRouteRoskas = route => route.legs.map(getRouteLegRow)
  |> stringsToTable
  |> #.toString()


export const printRoutes = (routes) => {
  if (!routes || !routes.length) {
    throw new Error('Cannot find any routes :(')
  }

  const rows = routes.reduce((acc, route) => [
    ...acc,
    getRouteHeaderRow(route),
    getRouteRoskas(route),
  ], [])

  console.log(rows.join('\n'), '\n')
  return rows
}
