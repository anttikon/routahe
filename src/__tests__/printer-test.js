import dummyRoute from './dummyRoute.json'
import { printRoutes, roundUpToNearest10 } from '../printer'

describe('printer', () => {
  describe('roundUpToNearest10', () => {
    it('should round up to nearest 10', () => {
      const data = [
        { from: 0, rounded: 0 },
        { from: 110.22, rounded: 110 },
        { from: 97.2, rounded: 100 },
        { from: 143.22, rounded: 150 },
        { from: 8.22, rounded: 10 }
      ]
      data.forEach(t => expect(roundUpToNearest10(t.from)).toEqual(t.rounded))
    })
  })

  describe('printRoutes', () => {
    it('should print routes', () => {
      const printedRows = printRoutes(dummyRoute)
      expect(printedRows.join('\n')).toMatchSnapshot()
    })
  })
})
