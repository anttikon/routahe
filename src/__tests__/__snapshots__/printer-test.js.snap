// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printer printRoutes should print routes 1`] = `
"
08:54 - 10:30 96min
[90m  |[39m 08:54 - 08:59 ðŸš¶ [90mWALK[39m   [90m340m[39m            -> Kamppi                          
[33m  |[39m 08:59 - 09:18 ðŸš‡ [33mSUBWAY[39m M1              -> MatinkylÃ¤                       
[90m  |[39m 09:18 - 09:21 ðŸš¶ [90mWALK[39m   [90m240m[39m            -> MatinkylÃ¤ (M)                   
[94m  |[39m 09:24 - 09:41 ðŸšŒ [94mBUS[39m    145             -> Suinonsalmi                     
[90m  |[39m 09:41 - 09:42 ðŸš¶ [90mWALK[39m   [90m140m[39m            -> Suinonsalmen saaristovenelaituri
[36m  |[39m 09:55 - 10:15 ðŸš¢ [36mFERRY[39m  Pentalan reitti -> Pentalan saaristovenelaituri    
[90m  |[39m 10:15 - 10:30 ðŸš¶ [90mWALK[39m   [90m1100m[39m           -> Destination                     

20:07 - 20:44 36min
[90m  |[39m 20:07 - 20:11 ðŸš¶ [90mWALK[39m   [90m190m[39m -> Pasila       
[35m  |[39m 20:11 - 20:16 ðŸš… [35mRAIL[39m   N    -> Helsinki     
[90m  |[39m 20:16 - 20:21 ðŸš¶ [90mWALK[39m   [90m440m[39m -> Rautatientori
[33m  |[39m 20:24 - 20:44 ðŸš‡ [33mSUBWAY[39m M2   -> MellunmÃ¤ki   
[90m  |[39m 20:44 - 20:44 ðŸš¶ [90mWALK[39m   [90m70m[39m  -> Destination  

20:17 - 20:54 37min
[90m  |[39m 20:17 - 20:19 ðŸš¶ [90mWALK[39m   [90m110m[39m -> Pasila       
[35m  |[39m 20:19 - 20:24 ðŸš… [35mRAIL[39m   Z    -> Helsinki     
[90m  |[39m 20:24 - 20:27 ðŸš¶ [90mWALK[39m   [90m270m[39m -> Rautatientori
[33m  |[39m 20:34 - 20:54 ðŸš‡ [33mSUBWAY[39m M2   -> MellunmÃ¤ki   
[90m  |[39m 20:54 - 20:54 ðŸš¶ [90mWALK[39m   [90m70m[39m  -> Destination  "
`;
